core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.typereg;
+
+import org.apache.brooklyn.api.internal.AbstractBrooklynObjectSpec;
+import org.apache.brooklyn.api.mgmt.ManagementContext;
+import org.apache.brooklyn.api.typereg.RegisteredType;
+import org.apache.brooklyn.api.typereg.RegisteredTypeLoadingContext;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.javalang.JavaClassNames;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Convenience supertype for {@link BrooklynTypePlanTransformer} instances.
+ */
+public abstract class AbstractTypePlanTransformer implements BrooklynTypePlanTransformer {
+
+    private static final Logger log = LoggerFactory.getLogger(AbstractTypePlanTransformer.class);
+    
+    protected ManagementContext mgmt;
+
+    @Override
+    public void injectManagementContext(ManagementContext mgmt) {
+        this.mgmt = mgmt;
+    }
+
+    private final String format;
+    private final String formatName;
+    private final String formatDescription;
+    
+    protected AbstractTypePlanTransformer(String format, String formatName, String formatDescription) {
+        this.format = format;
+        this.formatName = formatName;
+        this.formatDescription = formatDescription;
+    }
+    
+    @Override
+    public String getFormatCode() {
+        return format;
+    }
+
+    @Override
+    public String getFormatName() {
+        return formatName;
+    }
+
+    @Override
+    public String getFormatDescription() {
+        return formatDescription;
+    }
+
+    @Override
+    public String toString() {
+        return getFormatCode()+":"+JavaClassNames.simpleClassName(this);
+    }
+    
+    @Override
+    public double scoreForType(RegisteredType type, RegisteredTypeLoadingContext context) {
+        if (getFormatCode().equals(type.getPlan().getPlanFormat())) return 1;
+        if (type.getPlan().getPlanFormat()==null)
+            return scoreForNullFormat(type.getPlan().getPlanData(), type, context);
+        else
+            return scoreForNonmatchingNonnullFormat(type.getPlan().getPlanFormat(), type.getPlan().getPlanData(), type, context);
```

Comment:
(Personal preference ahead) I am on the fail-fast side of the spectrum so my thinking is that this kind of "make every last desperate attempt to produce something usable" approach just makes implementations more complex without actually working in the wild.